src/Stratis.Features.FederatedPeg/TargetChain/InputConsolidator.cs

```diff
@@ -0,0 +1,250 @@
+        {
+            RequestPartialTransactionPayload payload = new RequestPartialTransactionPayload(RequestPartialTransactionPayload.ConsolidationDepositId).AddPartial(this.PartialTransaction);
+            this.federatedPegBroadcaster.BroadcastAsync(payload).GetAwaiter().GetResult();
+        }
+
+        /// <inheritdoc />
+        public ConsolidationSignatureResult CombineSignatures(Transaction incomingPartialTransaction)
+        {
+            lock (this.lockObj)
+            {
+                // No need to sign in these cases.
+                if (!this.signingInProgress || this.fullySigned)
+                    return ConsolidationSignatureResult.Failed();
+
+                // Attempt to merge signatures
+                var builder = new TransactionBuilder(this.network);
+                Transaction oldTransaction = this.PartialTransaction;
+
+                this.logger.LogDebug("Attempting to merge signatures for {0} and {1}.", this.PartialTransaction.GetHash(), incomingPartialTransaction.GetHash());
+
+                this.PartialTransaction = SigningUtils.CombineSignatures(builder, this.PartialTransaction, new []{incomingPartialTransaction});
+
+                if (oldTransaction.GetHash() == this.PartialTransaction.GetHash())
+                {
+                    // Signing didn't work if the hash is still the same
+                    this.logger.LogDebug("Signing failed.");
+                    return ConsolidationSignatureResult.Failed();
+                }
+
+                this.logger.LogDebug("Successfully signed transaction.");
+
+                // NOTE: We don't need to reserve the transaction. The wallet will be at a standstill whilst this is happening.
+
+                // If it is FullySigned, broadcast.
+                if (this.walletManager.ValidateConsolidatingTransaction(this.PartialTransaction, true))
+                {
+                    this.logger.LogDebug("Consolidation transaction is fully signed. Broadcasting {0}", this.PartialTransaction.GetHash());
+                    this.broadcasterManager.BroadcastTransactionAsync(this.PartialTransaction);
+                    this.fullySigned = true;
+                }
+
+                return ConsolidationSignatureResult.Succeeded(this.PartialTransaction);
+            }
+        }
+
+        /// <summary>
+        /// Build a consolidating transaction.
+        /// </summary>
+        public Transaction BuildConsolidatingTransaction()
+        {
+            try
+            {
+                // TODO: Confifm that we can remove the ordering below.
+
+                List<UnspentOutputReference> unspentOutputs = this.walletManager.GetSpendableTransactionsInWallet(WithdrawalTransactionBuilder.MinConfirmations).ToList();
+
+                if (unspentOutputs.Count < WithdrawalTransactionBuilder.MaxInputs)
+                    throw new Exception("We shouldn't be consolidating transactions if we have less than 50 UTXOs to spend.");
```

Comment:
Should we throw an exception like this here? Why not just return null?

Reply:
Happy to change, see comment on try catch

Reply:
Yeah I reckon change this so that we log the values (`unspentOutputs.Count` and `WithdrawalTransactionBuilder.MaxInputs`) and then return null.